### PR TITLE
Fix OpenSSL v1.1 compatibility

### PR DESCRIPTION
**Pull request by hrobeers**
Runs and tests pass using openssl 1.1.0.f on Archlinux
Also tested for backwards compatibility with openssl 1.0

Please test if it compiles and runs on your system, heavy testing will be done on the second fork client.

### CODE REVIEW

**Comment by sunnyking** [comment]
Is this code change from Bitcoin, if so is it possible to reference a bitcoin change here? Also, is there impact to build requirements?

**Comment by hrobeers** [comment]
@sunnyking 
Bitcoin removed the OpenSSL depenency in v0.12. Therefore they did not have to upgrade to OpenSSL v1.1.
We figured that fixing the compatibility with OpenSSL v1.1 is much lower risk than doing another rebase to v0.12 under time pressure.

The fix should be very straightforward to review for anyone familiar with OpenSSL. In the mean time I tested it to fully validate both the mainnet & testnet chain.

**Comment by saeveritt** [comment]
Compiled on Debian 9.1  and running with OpenSSL 1.1.0f on testnet.

File: src/key.cpp [hunk 1]
```diff
@@ -354,8 +375,21 @@ bool CKey::SetCompactSignature(uint256 hash, const std::vector<unsigned char>& v
...
+    BN_bin2bn(&vchSig[1],32,sig_r);
+    BN_bin2bn(&vchSig[33],32,sig_s);
+
+    #if OPENSSL_VERSION_NUMBER > 0x1000ffffL
+    ECDSA_SIG_set0(sig, sig_r, sig_s);
+    #else
+    if (sig->r) BN_clear_free(sig->r);
+    if (sig->s) BN_clear_free(sig->s);
+    sig->r = sig_r;
+    sig->s = sig_s;
```
Comment (sigmike) [inline src/key.cpp]: This changes the behavior a bit. I don't know the internals of openssl, but if it ever keeps another reference to `r` and `s`  then this code breaks it. It's most likely not the case, but keeping the old code as it was would be simpler and less risky.

**Comment by sigmike** [comment]
The code looks good to me, except for all the new `*_new` function calls where allocation failure is not handled.

File: src/key.cpp (hunk 1 above)
Comment (hrobeers) [inline src/key.cpp]: I wanted to avoid duplication of the BN_bin2bn function calls.
But I agree keeping the old code here might be safer. Will change.

**Comment by hrobeers** [comment]
Thanks, @sigmike great feedback.
I forgot about the allocation failures indeed, will fix!

How did you think to handle a possible allocation failure in the Bignum ctor?
Throwing std::bad_alloc ?

**Comment by sigmike** [comment]
Yes.

**Comment by hrobeers** [comment]
I decided to reuse the bignum_error exception for consistency.

**Comment by hrobeers** [comment]
Ok this should be it guys.

Rewrote history squashed everything in one commit for clarity.

**Pull request merged**